diff --git a/src/nvim/CMakeLists.txt b/src/nvim/CMakeLists.txt
index 296a6d247f..f3d4d80962 100644
--- a/src/nvim/CMakeLists.txt
+++ b/src/nvim/CMakeLists.txt
@@ -530,7 +530,7 @@ if(CMAKE_CROSSCOMPILING AND NLUA0_HOST_PRG)
   set(LUA_GEN_DEPS ${GENERATOR_PRELOAD} ${NLUA0_HOST_PRG})
 else()
   set(LUA_GEN ${LUA_GEN_PRG} ${GENERATOR_PRELOAD} ${PROJECT_SOURCE_DIR} $<TARGET_FILE:nlua0> ${PROJECT_BINARY_DIR})
-  set(LUA_GEN_DEPS ${GENERATOR_PRELOAD} $<TARGET_FILE:nlua0>)
+  set(LUA_GEN_DEPS ${GENERATOR_PRELOAD} nlua0 $<TARGET_FILE:nlua0>)
 endif()
 
 # Like LUA_GEN but includes also vim.fn, vim.api, vim.uv, etc
